### ISSUE TITLE
error: static assertion failed with "incompatible 64-bit atomic layout"

### ISSUE THREAD

**Issue by ezyang** [facebookarchive/caffe2#273]
I am attempting to build caffe2 on the latest MacBook Pro. Here is some diagnostic information:

```
zcs:build ezyang$ clang --version
Apple LLVM version 8.0.0 (clang-800.0.38)
Target: x86_64-apple-darwin16.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
zcs:build ezyang$ /usr/local/cuda/bin/nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Tue_Jan_10_13:22:46_CST_2017
Cuda compilation tools, release 8.0, V8.0.61
```

I am failing with the following error:

```
[ 52%] Built target Caffe2_CPU
[ 53%] Building NVCC (Device) object caffe2/CMakeFiles/Caffe2_GPU.dir/utils/Caffe2_GPU_generated_math_gpu.cu.o
/usr/local/include/google/protobuf/stubs/atomicops_internals_generic_c11_atomic.h(52): error: static assertion failed with "incompatible 32-bit atomic layout"

/usr/local/include/google/protobuf/stubs/atomicops_internals_generic_c11_atomic.h(145): error: static assertion failed with "incompatible 64-bit atomic layout"

/Users/ezyang/Dev/caffe2/caffe2/utils/conversions.h(137): warning: variable "r" was declared but never referenced

2 errors detected in the compilation of "/var/folders/3t/cyx21zkx56qfp4633rbfcq00g9lvm8/T//tmpxft_00013a0d_00000000-19_math_gpu.compute_61.cpp1.ii".
CMake Error at Caffe2_GPU_generated_math_gpu.cu.o.Release.cmake:263 (message):
  Error generating file
  /Users/ezyang/Dev/caffe2/build/caffe2/CMakeFiles/Caffe2_GPU.dir/utils/./Caffe2_GPU_generated_math_gpu.cu.o


make[2]: *** [caffe2/CMakeFiles/Caffe2_GPU.dir/utils/Caffe2_GPU_generated_math_gpu.cu.o] Error 1

```

Google suggests that this is due to my protobufs being too old and not incorporating https://github.com/google/protobuf/pull/2699

However, what I find especially puzzling is that caffe2 comes with a vendored copy of protobufs in third_party, but as you can see from the make log, protobufs is being taken from /usr/local/header which is probably from brew install. So... how do I make caffe2 build with our vendored copy?

**Comment by Yangqing** [comment]
Yeah, protobuf is usually a very tricky issue to deal with due to the multiple installs. You might want to delete the brew protobuf library. Which version of protobuf do you have?

**Comment by victorv** [comment]
CUDA 8 is not compatible with Xcode 8 - see [here](http://docs.nvidia.com/cuda/cuda-installation-guide-mac-os-x/#axzz4ektfwCIS) - unless you are on Sierra.

You likely need to downgrade your Xcode to 7.3.x

**Comment by ezyang** [comment]
OK, I'm going to close this.

**Comment by codecolony** [comment]
Hi @ezyang ,
Could you solve this issue?

**Comment by victorv** [comment]
@codecolony - did you run into this problem?

The solution is to match you CUDA and Xcode tools properly.

**Comment by codecolony** [comment]
@victorv - Thank you. I'll try to do that. I had downgraded xcode to install some other framework. This is such a pain.

**Comment by robottomw** [comment]
I also saw this error with a compatible combination of XCode 8.2, macOS 10.12, LLVM 8.0.0 and CUDA 8.0.61. Turns out that it is also a known issue in protobuf 3.2.0 and seems to be fixed in protobuf 3.3.0:

https://github.com/caffe2/caffe2/issues/290
